File: iot/amqp/tlv/impl/AMQPError.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.ErrorCode as ErrorCode
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.wrappers.AMQPSymbol as AMQPSymbol
import iot.amqp.tlv.impl.TLVList as TLVList
import iot.amqp.tlv.impl.TLVFixed as TLVFixed

class amqpError():
    def __init__(self, condition, description, info):
        self.condition = condition
        self.description = description
        self.info = info

    def toArgumentsList(self):
        wrapper = AMQPWrapper.amqpWrapper()
        list = TLVList.tlvList(None,None)
        if self.condition is not None:
            list.addElement(0, wrapper.wrap(AMQPSymbol.amqpSymbol(self.condition)))
        if self.description is not None:
            list.addElement(1, wrapper.wrap(self.description))
        if self.info is not None:
            list.addElement(2, wrapper.wrap(self.info))

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(AMQPType.amqpType.getValueByKey('SMALL_ULONG'), 0x1D))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        if len(list.getList()) > 0 :
            element = list.getList()[0]
            if element is not None:
                self.condition = ErrorCode.errorCode(unwrapper.unwrapSymbol(element).getValue())
                print('AMQPError ERROR OCCURED condition= ' + str(ErrorCode.errorCode(unwrapper.unwrapSymbol(element).getValue())))
        if len(list.getList()) > 1 :
            element = list.getList()[1]
            if element is not None:
                self.description = unwrapper.unwrapString(element)
        if len(list.getList()) > 2 :
            element = list.getList()[2]
            if element is not None:
                self.info = unwrapper.unwrapMap(element)

    def toString(self):
        return 'AMQPError [condition='+ str(self.condition) + ', description=' + str(self.description) + ', info=' + str(self.info) + ']'

    def getCondition(self):
        return self.condition

    def setCondition(self, condition):
        self.condition = condition

    def getDescription(self):
        return self.description

    def setDescription(self, description):
        self.description = description

    def getInfo(self):
        return self.info

    def setInfo(self, info):
        self.info = info

    def isNull(self):
        return False

File: IoTbrokerClient/src/iot-protocols/amqp/classes/tlv/described/amqperror.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqperror.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"

AMQPError::AMQPError()
{
    this->condition = new AMQPErrorCode();
    this->descriptionString;
    this->info = map<JsonVariant *, JsonVariant *>();
}

void AMQPError::fill(AMQPTLVList *list)
{
    if (list->getList().size() > 0) {
        TLVAMQP *element = list->getList()[0];
        if (!element->isNull()) {
            String name = AMQPUnwrapper::unwrapSymbol(element)->getValue();
            this->condition = new AMQPErrorCode(this->condition->EnumObject::getValue(name));
        }
    }

    if (list->getList().size() > 1) {
        TLVAMQP *element = list->getList()[1];
        if (!element->isNull()) {
            this->descriptionString = AMQPUnwrapper::unwrapString(element);
        }
    }

    if (list->getList().size() > 2) {
        TLVAMQP *element = list->getList()[2];
        if (!element->isNull()) {
            this->info = AMQPUnwrapper::unwrapMap(element);
        }
    }
}

void AMQPError::addInfo(String key, JsonVariant *value)
{
    JsonVariant *keyVariant;
    keyVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(key)));
    this->info.insert(keyVariant, value);
}

AMQPTLVList *AMQPError::getList()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (this->condition != NULL) {
        JsonVariant *symbol;
        symbol.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(this->condition->getName())));
        list->addElementWithIndex(0, AMQPWrapper::wrapObject(symbol));
    }
    if (!this->descriptionString.isNull()) {
        list->addElementWithIndex(1, AMQPWrapper::wrapString(this->descriptionString));
    }
    if (this->info.count() != 0) {
        list->addElementWithIndex(2, AMQPWrapper::wrapMap(this->info));
    }

    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);

    unsigned char * data;
    data = malloc(sizeof(char));
    data[0] = (char) 0x1D;

    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);
    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);
    list->setConstructor(constructor);

    return list;
}

AMQPErrorCode *AMQPError::getCondition() const
{
    return condition;
}

void AMQPError::setCondition(AMQPErrorCode *value)
{
    condition = value;
}

String AMQPError::getDescriptionString() const
{
    return descriptionString;
}

void AMQPError::setDescriptionString(const String &value)
{
    descriptionString = value;
}

map<JsonVariant *, JsonVariant *> AMQPError::getInfo() const
{
    return info;
}
